File: alamos.go
package main

import (
	"fmt"
	"os"
	"slices"
)

const (
	HIGH_VALUE   = 100000
	SEARCH_DEPTH = 6
)

const (
	EMPTY = 0
	wP    = 1
	bP    = -1
	wN    = 2
	bN    = -2
	wR    = 3
	bR    = -3
	wQ    = 4
	bQ    = -4
	wK    = 5
	bK    = -5
)

func evaluateBoard(board []int) int {
	pieceValues := []int{
		-10000, -900, -500, -300, -100,
		0,
		100, 300, 500, 900, 10000,
	}
	score := 0
	for _, piece := range board {
		score += pieceValues[piece+5]
	}

	return score
}

func offsetMoves(position [2]int, board []int, offsets [][2]int) [][2]int {
	var moves [][2]int
	piece := board[position[0]*6+position[1]]
	for _, offset := range offsets {
		nx, ny := position[0]+offset[0], position[1]+offset[1]
		if nx >= 0 && nx < 6 && ny >= 0 && ny < 6 {
			target := board[nx*6+ny]
			if target == EMPTY || target*piece < 0 {
				moves = append(moves, [2]int{nx, ny})
			}
		}
	}
	return moves
}

func getKnightMoves(position [2]int, board []int) [][2]int {
	offsets := [][2]int{{2, 1}, {1, 2}, {-2, 1}, {-1, 2}, {2, -1}, {1, -2}, {-2, -1}, {-1, -2}}
	return offsetMoves(position, board, offsets)
}

func getKingMoves(position [2]int, board []int) [][2]int {
	offsets := [][2]int{{1, 0}, {0, 1}, {-1, 0}, {0, -1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}}
	return offsetMoves(position, board, offsets)
}

func vectorMoves(position [2]int, board []int, vectors [][2]int) [][2]int {
	var moves [][2]int
	piece := board[position[0]*6+position[1]]
	for _, vector := range vectors {
		nx, ny := position[0], position[1]
		for {
			nx, ny = nx+vector[0], ny+vector[1]
			if nx >= 0 && nx < 6 && ny >= 0 && ny < 6 {
				target := board[nx*6+ny]
				if target == EMPTY {
					moves = append(moves, [2]int{nx, ny})
					continue
				}
				if target*piece < 0 {
					moves = append(moves, [2]int{nx, ny})
				}
			}
			break
		}
	}
	return moves
}

func getRookMoves(position [2]int, board []int) [][2]int {
	vectors := [][2]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
	return vectorMoves(position, board, vectors)
}

func getQueenMoves(position [2]int, board []int) [][2]int {
	vectors := [][2]int{{1, 1}, {1, -1}, {-1, 1}, {-1, -1}}
	var moves [][2]int = vectorMoves(position, board, vectors)
	moves = append(moves, getRookMoves(position, board)...)
	return moves
}

func getPawnMoves(position [2]int, board []int) [][2]int {
	var moves [][2]int
	x, y := position[0], position[1]
	piece := board[x*6+y]

	dx := 1
	if piece > 0 {
		dx = -1
	}

	if x+dx >= 0 && x+dx < 6 {
		if board[(x+dx)*6+y] == EMPTY {
			moves = append(moves, [2]int{x + dx, y})
		}

		for _, dy := range []int{-1, 1} {
			if y+dy >= 0 && y+dy < 6 {
				target := board[(x+dx)*6+y+dy]
				if target != EMPTY && target*piece < 0 {
					moves = append(moves, [2]int{x + dx, y + dy})
				}
			}
		}
	}

	return moves
}

func makeMove(board []int, position [2]int, move [2]int) []int {
	posIdx := position[0]*6 + position[1]
	moveIdx := move[0]*6 + move[1]
	piece := board[posIdx]

	newBoard := make([]int, len(board))
	copy(newBoard, board)

	newBoard[posIdx] = EMPTY
	newBoard[moveIdx] = piece

	if piece == wP && move[0] == 5 {
		newBoard[moveIdx] = wQ
	} else if piece == bP && move[0] == 0 {
		newBoard[moveIdx] = bQ
	}

	return newBoard
}

func getMoves(idx int, piece int, board []int) [][2]int {
	var moves [][2]int
	position := [2]int{idx / 6, idx % 6}

	switch piece {
	case wP, bP:
		moves = getPawnMoves(position, board)
	case wR, bR:
		moves = getRookMoves(position, board)
	case wN, bN:
		moves = getKnightMoves(position, board)
	case wQ, bQ:
		moves = getQueenMoves(position, board)
	case wK, bK:
		moves = getKingMoves(position, board)
	}

	return moves
}

func negamax(board []int, depth, alpha, beta, color int) ([]int, int) {
	if depth == 0 || !slices.Contains(board, wK) || !slices.Contains(board, bK) {
		return board, color * evaluateBoard(board)
	}

	bestValue := -HIGH_VALUE
	bestBoard := board

	for idx, piece := range board {
		if (color == 1 && piece > 0) || (color == -1 && piece < 0) {
			moves := getMoves(idx, piece, board)
			for _, move := range moves {
				newBoard := makeMove(board, [2]int{idx / 6, idx % 6}, move)
				_, moveValue := negamax(newBoard, depth-1, -beta, -alpha, -color)
				moveValue = -moveValue

				if moveValue > bestValue {
					bestValue = moveValue
					bestBoard = newBoard
				}

				if bestValue > alpha {
					alpha = bestValue
				}
				if alpha >= beta {
					break
				}
			}
		}
		if alpha >= beta {
			break
		}
	}
	return bestBoard, bestValue
}

var charToIntMap = map[rune]int{
	'.': EMPTY,
	'P': wP,
	'p': bP,
	'R': wR,
	'r': bR,
	'N': wN,
	'n': bN,
	'Q': wQ,
	'q': bQ,
	'K': wK,
	'k': bK,
}

func stringToIntBoard(board string) []int {
	intBoard := make([]int, 36)
	for i, piece := range board {
		intBoard[i] = charToIntMap[piece]
	}
	return intBoard
}

func intToStringBoard(intBoard []int) string {
	board := make([]rune, 36)
	for i, piece := range intBoard {
		for char, val := range charToIntMap {
			if val == piece {
				board[i] = char
				break
			}
		}
	}
	return string(board)
}

func main() {
	inputBoard := stringToIntBoard(os.Args[1])
	color := 1
	if os.Args[2] == "b" {
		color = -1
	}

	board, _ := negamax(inputBoard, SEARCH_DEPTH, -HIGH_VALUE, HIGH_VALUE, color)

	nextTurn := "w"
	if os.Args[2] == "w" {
		nextTurn = "b"
	}
	fmt.Println(intToStringBoard(board), nextTurn)
}

File: alamos.lua
#!/usr/bin/lua

HIGH_VALUE = 100000
SEARCH_DEPTH = 6

EMPTY = 0
wP = 1
bP = -1
wN = 2
bN = -2
wR = 3
bR = -3
wQ = 4
bQ = -4
wK = 5
bK = -5

function table_contains(table, value)
  for _, v in ipairs(table) do
    if v == value then
      return true
    end
  end
  return false
end

function evaluate_board(board)
  local piece_values = {-10000, -900, -500, -300, -100, 0, 100, 300, 500, 900, 10000}
  local sum = 0
  for _, piece in ipairs(board) do
    sum = sum + piece_values[piece + 6]
  end
  return sum
end

function offset_moves(position, board, offsets)
  local results = {}
  for _, offset in ipairs(offsets) do
    local dx, dy = offset[1], offset[2]
    local nx, ny = position[1] + dx, position[2] + dy
    if nx >= 1 and nx <= 6 and ny >= 1 and ny <= 6 then
      local target = board[(nx - 1) * 6 + ny]
      if target * board[(position[1] - 1) * 6 + position[2]] <= 0 then
        table.insert(results, {nx, ny})
      end
    end
  end
  return results
end

function get_knight_moves(position, board)
  local offsets = {{2, 1}, {1, 2}, {-2, 1}, {-1, 2}, {2, -1}, {1, -2}, {-2, -1}, {-1, -2}}
  return offset_moves(position, board, offsets)
end

function get_king_moves(position, board)
  local offsets = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}}
  return offset_moves(position, board, offsets)
end

function vector_moves(position, board, vectors)
  local results = {}
  for _, vector in ipairs(vectors) do
    local dx, dy = vector[1], vector[2]
    local nx, ny = position[1], position[2]
    while true do
      nx, ny = nx + dx, ny + dy
      if nx >= 1 and nx <= 6 and ny >= 1 and ny <= 6 then
        local target = board[(nx - 1) * 6 + ny]
        if target == EMPTY then
          table.insert(results, {nx, ny})
        else
          if target * board[(position[1] - 1) * 6 + position[2]] < 0 then
            table.insert(results, {nx, ny})
          end
          break
        end
      else
        break
      end
    end
  end
  return results
end

function get_rook_moves(position, board)
  local vectors = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
  return vector_moves(position, board, vectors)
end

function get_queen_moves(position, board)
  local results = {}
  local vectors = {{1, 1}, {1, -1}, {-1, 1}, {-1, -1}}
  for _, move in ipairs(vector_moves(position, board, vectors)) do
    table.insert(results, move)
  end
  for _, move in ipairs(get_rook_moves(position, board)) do
    table.insert(results, move)
  end
  return results
end

function get_pawn_moves(position, board)
  local x, y = position[1], position[2]
  local dx = board[(x - 1) * 6 + y] < 0 and 1 or -1
  local results = {}

  if x + dx >= 1 and x + dx <= 6 then
    if board[(x + dx - 1) * 6 + y] == EMPTY then
      table.insert(results, {x + dx, y})
    end
    for _, dy in ipairs({-1, 1}) do
      if y + dy >= 1 and y + dy <= 6 then
        local target = board[(x + dx - 1) * 6 + y + dy]
        if target * board[(x - 1) * 6 + y] < 0 then
          table.insert(results, {x + dx, y + dy})
        end
      end
    end
  end
  return results
end

function make_move(board, position, move)
  local new_board = {}
  for i, v in ipairs(board) do
    new_board[i] = v
  end
  local pos_idx = (position[1] - 1) * 6 + position[2]
  local move_idx = (move[1] - 1) * 6 + move[2]
  local piece = new_board[pos_idx]

  new_board[pos_idx] = EMPTY
  new_board[move_idx] = piece

  if piece == wP and move[1] == 5 then
    new_board[move_idx] = wQ
  elseif piece == bP and move[1] == 0 then
    new_board[move_idx] = bQ
  end

  return new_board
end

function get_moves(idx, piece, board)
  local x = math.floor((idx - 1) / 6) + 1
  local y = (idx - 1) % 6 + 1
  local moves = {}

  local piece_sets = {
    {wP, bP},
    {wR, bR},
    {wN, bN},
    {wQ, bQ},
    {wK, bK}
  }

  local funcs = {
    get_pawn_moves,
    get_rook_moves,
    get_knight_moves,
    get_queen_moves,
    get_king_moves
  }

  for i, set in ipairs(piece_sets) do
    for _, p in ipairs(set) do
      if piece == p then
        for _, move in ipairs(funcs[i]({x, y}, board)) do
          table.insert(moves, move)
        end
        return moves
      end
    end
  end

  return moves
end

function negamax(board, depth, alpha, beta, color)
  if depth == 0 or not table_contains(board, wK) or not table_contains(board, bK) then
    return board, color * evaluate_board(board)
  end

  local best_value = -HIGH_VALUE
  local best_board = board

  for idx, piece in ipairs(board) do
    if (color == 1 and piece > 0) or (color == -1 and piece < 0) then
      for _, move in ipairs(get_moves(idx, piece, board)) do
        local new_board = make_move(board, {math.floor((idx - 1) / 6) + 1, (idx - 1) % 6 + 1}, move)
        local _, move_value = negamax(new_board, depth - 1, -beta, -alpha, -color)
        move_value = -move_value
        if move_value > best_value then
          best_value = move_value
          best_board = new_board
        end
        alpha = math.max(alpha, best_value)
        if alpha >= beta then
          break
        end
      end
      if alpha >= beta then
        break
      end
    end
  end

  return best_board, best_value
end

char_to_int_map = {
  ["."] = EMPTY,
  ["P"] = wP,
  ["p"] = bP,
  ["R"] = wR,
  ["r"] = bR,
  ["N"] = wN,
  ["n"] = bN,
  ["Q"] = wQ,
  ["q"] = bQ,
  ["K"] = wK,
  ["k"] = bK
}

function string_to_int_board(board_string)
  local board = {}
  for i = 1, #board_string do
    local c = board_string:sub(i,i)
    table.insert(board, char_to_int_map[c])
  end
  return board
end

function int_to_string_board(int_board)
  local board_string = ""
  for _, piece in ipairs(int_board) do
    for char, int_value in pairs(char_to_int_map) do
      if int_value == piece then
        board_string = board_string .. char
        break
      end
    end
  end
  return board_string
end

-- Main execution
local args = {...}

local board, _ = negamax(
  string_to_int_board(args[1]),
  SEARCH_DEPTH,
  -HIGH_VALUE,
  HIGH_VALUE,
  (args[2] == "w" and 1) or -1
)

print(int_to_string_board(board) .. " " .. ((args[2] == "w" and "b") or "w"))
